[Pallas TPU] Add Pipelining and BlockSpecs documentation for Pallas TPU (#20001)
google/jax: docs/pallas/tpu/pipelining.md

* **Registers**: A TensorCore has two main types of registers: vector registers (VREGs) store array values, and scalar registers (SREGs) store scalar values. Values can be loaded into memory from their respective caches (VMEM for VREGs and SMEM for SREGs).
* **Compute units**: A TensorCore has a scalar unit, vector unit (VPU) and matrix unit (MXU) that can do numerical computation. Compute units operate on values that live in SREGs and VREGs and output values into those registers as well.

In order to do a vectorized computation on our values `x` and `y` that live in HBM, we need to:

1. Copy the values `x` and `y` into VMEM.
2. Load the values from VMEM into VREGs.
3. Execute the computation using the VPU or MXU, storing the output in VREGs.
4. Store the values in the output VREGs into VMEM.
5. Copy the output values in VMEM back to HBM.

+++ {"id": "TzctMbNsn3vc"}

Let's implement a Pallas function that does just that!

```{code-cell}
:id: 2IXQxNWrKJyb
:outputId: d62eb493-5f92-4496-f113-d3cd24cb0b9f

def add_matrices_kernel(x_vmem_ref, y_vmem_ref, z_vmem_ref):
  # Load x and y from VMEM into VREGs

>   x_vregs = x_vmem_ref[...]

Reviewer: Maybe use `[:, :]` instead of the rank-polymorphic `[...]`, given the name of the function?

  y_vregs = y_vmem_ref[...]
  # Execute a vectorized add
  z_vregs = x_vregs + y_vregs
  # Store the output values in VREGs back into VMEM
  z_vmem_ref[...] = z_vregs

def add_matrices(x: jax.Array, y: jax.Array) -> jax.Array:

>   # pallas_call will (by default) copy `x` and `y` from HBM into VMEM. It will

Reviewer: Is saying "it will also allocate VMEM buffers for input" redundant with "pallas_call will copy `x` and `y` from HBM into VMEM"?

Reviewer: Maybe the second sentence here meant to say "intermediates"?

+++ {"id": "5kWr-1tKpYro"}

## Constraints of using VMEM/SMEM

Pallas exposes access to lower level memory spaces like VMEM and SMEM but writing kernels utilizing them adds some considerations.

1. Memory capacity. VMEM and SMEM are *small*! VMEM on v4 TPUs is only 16MiB and SMEM ranges in the tens to hundreds of KiB. If our arrays are too big, we won't even be able to fit them into VMEM at all. For reference, a `f32[2048, 2048]` array is 16MiB, so our above kernel won't scale beyond moderately sized arrays.

2. Memory bandwidth. Copying to/from HBM and VMEM takes a long time, at least compared to most compute instructions. The `add_matrices` function above will likely spend more time copying between HBM and VMEM than actually performing the addition itself.

> With these two constraints in mind, we'll have to rethink our strategy for getting performance of our TPUs.

Reviewer: ```suggestion
With these two constraints in mind, we'll have to rethink our strategy for getting performance out of our TPUs.
```

+++ {"id": "_NTqvlbetB3P"}

## Primer: Pipelining

Pipelining our computation offers a way of dealing with both the memory capacity and bandwidth constraints in one fell swoop. What do we mean by pipelining?

> The goal is: *concurrently* copy to/from HBM and VMEM *while* utilizing our compute units. However, the way we expressed our program above is that we copy *all* of `x` and `y` before we start doing any compute with them. This creates a dependence between the copy and the compute.

Reviewer: Might be slightly clearer to say 'in parallel' rather than 'concurrently', since we need these things to be physically running at the same time (and overlapping) not just logically separated (e.g. https://stackoverflow.com/a/1050257).

> The goal is: *concurrently* copy to/from HBM and VMEM *while* utilizing our compute units. However, the way we expressed our program above is that we copy *all* of `x` and `y` before we start doing any compute with them. This creates a dependence between the copy and the compute.

Reviewer: Maybe put the "However, the way we expressed ..." stuff in the preceding section? ATM it feels like we're jumping back to discussing the old kernel more right when we were starting to introduce pipelining. But if we move these sentences to the preceding section, we can in that section explain what's wrong with the previous way of writing things, and in this section focus on pipelining.

Author: This one is tough to move up because I'm trying to explain why we can't naively pipeline our code above because of the dependence, to motivate chunking. I reworded it a bit but I'll keep it here.

However, if we can chunk up our computation into several subcomputations (e.g. when we add two matrices, we can express that as addition of "blocks" of the original matrices together), we can now overlap the copies of one of those subcomputations with the compute of the other. Let's walk through a simple example:

> Let's say we split our arrays `x` and `y` into `x1, x2` and `y1, y2` (for example, split along the leading axis, resulting in two `(256, 512)` arrays for each input. We can now execute the following pipelined computation.

Reviewer: Maybe we can make a pair of timeline sketches illustrating how things look with and without pipelining for this example? (Or even better than a drawing would be if we could 'take a photo' of real timings, but I'm not sure if that's feasible.)

Author: Anselm pointed me to how to make animations of this, so I'll try that and add it as a follow up!

1. Copy `x1` and `y1` into VMEM.
1. Start copying `x2` and `y2` into VMEM
2. Load `x1, y1` from VMEM into VREGs.
3. Execute the `z1 = x1 + y1` using the compute units.
4. Store `z1` into VMEM.
5. Start copying `z1` from VMEM back into HBM.
6. Wait until `x2, y2` have been copied into VMEM.
7. Load `x2, y2` from VMEM into VREGs.
8. Execute the `z2 = x2 + y2` using the compute units.
9. Store `z2` into VMEM.
10. Wait until `z1` is copied into HBM.
10. Start copying `z2` from VMEM back into HBM.
10. Wait until `z2` is copied into HBM.

Any time we are doing compute here, we are asynchronously copying something. This means that some of the time spent copying is not wasted.

> The two most important numbers for determining how efficient a pipelined computation are a) how many FLOP/s we can execute and b) how many bytes/s we can copy. The ratio of these two (FLOPs/bandwidth) is called the *arithmetic intensity* of an operation.

Reviewer: I don't think this is consistent with [the wikipedia definition of arithmetic intensity](https://en.wikipedia.org/wiki/Roofline_model#:~:text=Arithmetic%20intensity,-The%20arithmetic%20intensity&text=will%20be%20the%20ratio%20of,movement%20(FLOPs%2Fbyte).).  That definition is about FLOPs (ie total number of floating point operations performed, not FLOPS or FLOP/s, ie floating point operations per second) compared to total bytes of traffic (ie not bytes/sec). Also I think those quantities are properties of a workload, e.g. matrix multiplication of particular shapes, rather than a chip's FLOP/s and bytes/s. See also [this ref](https://people.eecs.berkeley.edu/~kubitron/cs252/handouts/papers/RooflineVyNoYellow.pdf).

Maybe there are some other definitions of the term 'arithmetic intensity' that are consistent with what's written here, but if so can you link them? If not, let's adapt to be consistent with Wikipedia and link that. Or just delete this paragraph because I don't think arithmetic intensity is ever referenced again in the doc.

Author: Fixed.

+++ {"id": "gutx7y8uvZKH"}

## Pipelining in Pallas

+++ {"id": "U-dPTjlBverB"}

How do we implement a pipeline like the one above in Pallas? It seems like a complex sequence of asynchronous data operations and executing kernels that would be a pain to implement manually. Fear not! Pallas offers an API for expressing pipelines without too much boilerplate, namely through `grid`s and `BlockSpec`s.

+++ {"id": "x-LQKu8HwED7"}

### `grid`, a.k.a. kernels in a loop

See how in the above pipelined example, we are executing the same logic multiple times: steps 3-5 and 8-10 both execute the same operations, only on different inputs. We can imagine the generalized version of this is a loop in which the same kernel is executed multiple times.

The number of iterations in the loop is specified via the `grid` argument to `pallas_call`. Conceptually:
```python
pl.pallas_call(some_kernel, grid=n)(...)
```
maps to
```python
for i in range(n):
  # do HBM<->VMEM copies
  some_kernel(...)

>   # do HBM<->VMEM copies

Reviewer: Should the first one be `HBM -> VMEM` and the second one be `VMEM -> HBM`?

+++ {"id": "miWgPkytyIIa"}

The next piece of information we need to provide Pallas in order to automatically pipeline our computation is information on how to chunk it up. Specifically, we need to provide a mapping between *the iteration of the loop* to *which block of our inputs and outputs to be operated on*. A `BlockSpec` is exactly these two pieces of information.

 First we pick a `block_shape` for our inputs. In the pipelining example above, we had `(512, 512)`-shaped arrays and split them along the leading dimension into two `(256, 512)`-shaped arrays. In this pipeline, our `block_shape` would be `(256, 512)`.

We then provide an `index_map` function that maps the iteration space to the blocks. Specifically, in the aforementioned pipeline, on the 1st iteration we'd like to select `x1` and on the second iteration we'd like to use `x2`. This can be expressed with the following `index_map`:

```python
def x_index_map(i):
  return (i, 0)
```

We'd then construct the `BlockSpec`:
```python
x_block_spec = pl.BlockSpec(x_index_map, (256, 512))
```

The `y_block_spec` and `z_block_spec` will be identical to `x_block_spec`.

+++ {"id": "noybOKghzjwG"}

### Putting it together

> We provide these arguments to `pallas_call` via `grid`, `in_specs` and `out_specs` (`in_specs` correspond to inputs, `out_spec` correspond to outputs).

Reviewer: ```suggestion
We provide these arguments to `pallas_call` via `grid`, `in_specs` and `out_specs` (`in_specs` corresponds to the tuple of positional arguments, and `out_spec` corresponds to the output).
```

      out_specs=block_spec,
      grid=(2,))(x, y)

add_matrices_pipelined(x, y)
```

+++ {"id": "rkytgIZYzz4t"}

We've only added a little bit of code to our original function to add automatic pipelining but the `BlockSpec`s and `grid` do a lot of heavy lifting!

How does it work? Well, the `BlockSpec`s provide enough information to start *prefetching* blocks of our input from HBM into VMEM. For example, if we are starting iteration `i` of our `grid`, we can pass `i + 1` into the `index_map` functions to obtain the blocks needed for the next iteration. We can then start an asynchronous copy for those blocks. Similarly for outputs, we can wait for the outputs of the previous iteration to be copied before starting the copy for the current iteration's outputs.

+++ {"id": "7Xtz9oMs0ZRL"}

### Parameterizing a pipeline

+++ {"id": "esY4GcIB0bqQ"}

It's common to parameterize the block shapes in our kernel. Block sizes are perhaps the most important parameter to tune when optimizing the performance of Pallas kernels! They us more control over the pipeline (for example, picking smaller blocks adds more iterations to our pipelined loop where each iteration has less work to do).

> Furthermore, we could also carve up the inputs and outputs along the 2nd dimension (we are only splitting along the first right now). Let's write more general kernel that handles both of these additions.

Reviewer: ```suggestion
Furthermore, we could also carve up the inputs and outputs along the 2nd dimension (we are only splitting along the first right now). Let's write a more general kernel that handles both of these additions.
```

def sum(x: jax.Array) -> jax.Array:
  grid, *out_shape = x.shape
  return pl.pallas_call(
      sum_kernel,
      grid=grid,
      # None in `block_shape` means we pick a size of 1 and squeeze it away
      in_specs=[pl.BlockSpec(lambda i: (i, 0, 0), (None, *out_shape))],
      out_specs=pl.BlockSpec(lambda i: (0, 0), out_shape),
      out_shape=jax.ShapeDtypeStruct(out_shape, x.dtype)
      )(x)
sum(x)
```

+++ {"id": "2828qXBI5ksZ"}

This `sum` function now outputs the correct values!

One last thing to note about reductions in Pallas are that **they must be done in the minormost (rightmost) dimensions of our grid** (our grid is 1-dimensional in the above example so we are reducing over its minormost dimension). This is because the pipeline that Pallas generates using the `BlockSpec`s, `grid` and kernel function *does not read outputs back from HBM*. Once you've written an output value back to HBM you cannot revisit it. Therefore, you cannot do a reduction across a grid dimension that has any revisiting and therefore all reductions need to happen in the rightmost dimensions.

+++ {"id": "KvPFez9N8cKJ"}

## TPUs in Megacore configuration

+++ {"id": "0f4HAVzQ8n71"}

> Some TPU chips have two TensorCores but appear as one device to JAX users. This is what is called "megacore". The separate TensorCores have their own separate VMEM, VREGs, SMEM, SREGs and compute units but *share HBM*.

Reviewer: ```suggestion
Some TPU chips have two TensorCores but appear as one device to JAX users. This is called "megacore". The separate TensorCores have their own separate VMEM, VREGs, SMEM, SREGs and compute units but *share HBM*.
```

+++ {"id": "x-LQKu8HwED7"}

### `grid`, a.k.a. kernels in a loop

See how in the above pipelined example, we are executing the same logic multiple times: steps 3-5 and 8-10 both execute the same operations, only on different inputs. We can imagine the generalized version of this is a loop in which the same kernel is executed multiple times.

The number of iterations in the loop is specified via the `grid` argument to `pallas_call`. Conceptually:
```python
pl.pallas_call(some_kernel, grid=n)(...)
```
maps to
```python
for i in range(n):
  # do HBM<->VMEM copies
  some_kernel(...)
  # do HBM<->VMEM copies
```
Grids can be generalized to be multi-dimensional, corresponding to nested loops. For example,

```python
pl.pallas_call(some_kernel, grid=(n, m))(...)
```
is equivalent to
```python
for i in range(n):
  for j in range(m):
    # do HBM<->VMEM copies
    some_kernel(...)
    # do HBM<->VMEM copies
```
This generalizes to any tuple of integers (a length `d` grid will correspond to `d` nested loops).

+++ {"id": "hRLr5JeyyEwM"}

### `BlockSpec`, a.k.a. how to chunk up inputs

+++ {"id": "miWgPkytyIIa"}

> The next piece of information we need to provide Pallas in order to automatically pipeline our computation is information on how to chunk it up. Specifically, we need to provide a mapping between *the iteratio of the loop* to *which block of our inputs and outputs to be operated on*. A `BlockSpec` is exactly these two pieces of information.

Reviewer 2: ```suggestion
The next piece of information we need to provide Pallas in order to automatically pipeline our computation is information on how to chunk it up. Specifically, we need to provide a mapping between *the iteration of the loop* to *which block of our inputs and outputs to be operated on*. A `BlockSpec` is exactly these two pieces of information.
```

Therefore, **whenever we do a reduction in a kernel, we need to make sure to initialize the `Ref` that is storing the reduced value**. We can accomplish this using `pl.when` (which executes a function conditionally) and `pl.program_id` (which queries which iteration of the grid inside of a kernel).

```{code-cell}
---
executionInfo:
  elapsed: 91
  status: ok
  timestamp: 1708910182754
  user:
    displayName: Sharad Vikram
    userId: '16113763449895047367'
  user_tz: 480
id: JXN2RthX5cSw
outputId: 195df19b-a889-479b-95b6-1fb7281f1518
---
def sum_kernel(x_ref, o_ref):
  @pl.when(pl.program_id(0) == 0)
  def _():
    o_ref[...] = jnp.zeros_like(o_ref)

  o_ref[...] += x_ref[...]

def sum(x: jax.Array) -> jax.Array:
  grid, *out_shape = x.shape
  return pl.pallas_call(
      sum_kernel,
      grid=grid,
      # None in `block_shape` means we pick a size of 1 and squeeze it away
      in_specs=[pl.BlockSpec(lambda i: (i, 0, 0), (None, *out_shape))],
      out_specs=pl.BlockSpec(lambda i: (0, 0), out_shape),
      out_shape=jax.ShapeDtypeStruct(out_shape, x.dtype)
      )(x)
sum(x)
```

+++ {"id": "2828qXBI5ksZ"}

This `sum` function now outputs the correct values!

> One last thing to note about reductions in Pallas are that **they must be done in the minormost (rightmost) dimensions of our grid**. This is because the pipeline that Pallas generates using the `BlockSpec`s, `grid` and kernel function *does not read outputs back from HBM*. Once you've written an output value back to HBM you cannot revisit it. Therefore, you cannot do a reduction across a grid dimension that has any revisiting and therefore all reductions need to happen in the rightmost dimensions.

Reviewer 2: You say to reduce on the rightmost dimension but above you are reducing on the leftmost (major) dimension right?

Maybe the example should be changed?

Author: The `grid` is only one dimensional so technically we are reducing over its minormost dimension (not the minormost dimension of the inputs/outputs, it's the grid that matters). I'll add clarification.

Reviewer 2: Ahh, makes sense. Thanks!
